Committing with all proposed low-level changes (Uptane Standards Forum) (#110)
uptane/uptane-standard: uptane-standard.md

## Terminology

*Bundle*: A set of images released by the repository that is meant to be installed by one or more target ECUs on a vehicle during the same update cycle.

*Bus*: An internal communications network that interconnects components within a vehicle. A vehicle can have a number of buses that will vary in terms of power, speed and resources.

> *ECU Identifier*: A characteristic used to identify a specific ECU (e.g., such as a unique serial number).

Reviewer: ```suggestion
*ECU Identifier*: An attribute used to identify a specific ECU (e.g., such as a unique serial number).
```

> *ECU Version Manifest*: A file which details the software version currently installed on the ECU.

Reviewer: ```suggestion
*ECU Version Manifest*: Metadata which details the software version currently installed on the ECU.
```

> *Hardware Identifier*: A characteristic used to identify a model of an ECU.

Reviewer: ```suggestion
*Hardware Identifier*: An attribute used to identify a model of an ECU.
```

*ECUs*: Electronic Control Units, the computing units on a vehicle

*LIN Bus*: Local Interconnect Bus 

> *RXSWIN*: RX Software Identification Number, defined in UN WP29 as "a dedicated identifier, defined by the vehicle manufacturer, representing information about the type approval relevant software of the Electronic Control System contributing to the Regulation N*X type approval relevant characteristics of the vehicle." 

Reviewer 2: Why do we want to specify this if it is completely optional?

Reviewer: Agreed

Author: I believe it is helpful for implementer's to understand where international standards (UN WP 29 - utilizing RXSWIN) can be included in the metadata.

Please let me know if you do not believe this is helpful.

Author: I will remove, with proposal of including in Implementation Guide/Considerations.

#### Updating one ECU with a complete image

> A tier-1 supplier completes work on a revised image for an electronic brake control module. This module will control the brakes on all models of an SUV produced by the OEM mentioned above. Assuming supplier delegation is supported by the OEM for this ECU, each tier-1 supplier digitally signs the image, then delivers the signature and all of its metadata, including delegations, and associated images to the OEM. The OEM adds these metadata and images to its image repository, along with information about any dependencies and conflicts between this image and those on other ECUs used in the OEM's vehicles. The OEM also updates the inventory database, so that the director repository can instruct the ECU on how to install these updated images.

Reviewer: ```suggestion
A tier-1 supplier completes work on a revised image for an electronic brake control module. This module will control the brakes on all models of an SUV produced by the OEM mentioned above. Assuming supplier delegation is supported by the OEM for this ECU, each tier-1 supplier digitally signs the image, then delivers the signature and all of its metadata, including delegations, and associated images to the OEM. The OEM adds these metadata and images to its image repository, along with information about any dependencies and conflicts between this image and those for other ECUs used in the OEM's vehicles. The OEM also updates the inventory database, so that the director repository can instruct the ECU on how to install these updated images.
```

A repository's Root metadata distributes the public keys of the top-level Root, Targets, Snapshot, and Timestamp roles, as well as revocations of those keys. It SHALL contain two attributes:

> * A representation of the public keys for all four roles. Each key shall have a unique public key identifier.

Reviewer 2: Good catch, but then `SHALL` should be capitalized.

Author: done.

> * A representation of the public keys for all four roles. Each key shall have a unique public key identifier.

Reviewer: ```suggestion
* A representation of the public keys for all four roles. Each key SHALL have a unique public key identifier.
```

* The size of the image in bytes
* One or more hashes of the image file, along with the hashing function used

> This list MAY also include an RXSWIN identifier.

Reviewer: Should be removed, I think, like @patrickvacek noted

Author: That is fair, I will remove throughout the document.
I would like for RXSWIN to be included in the implementation guide/considerations document for clear guidance on how international standards (UN WP 29) may fit within Uptane.

The following information is CONDITIONALLY REQUIRED for each image on the Director repository IF that image is encrypted:

> 

Reviewer 2: Why was this removed?

Author: I removed it for clarity since it is already stated as a requirement in Section 5.2.3.1 "Metadata about Images".

Please let me know your feedback.

Author: Considering the ongoing conversation about primary's not having ability to decrypt all images, I have added this back.

Sorry for any confusion.

A Director repository MUST conform to the following six-step process for directing the installation of software images on a vehicle.

> 1. The Director must first identify the vehicle. This MAY be done when the Director receives a vehicle version manifest sent by a primary (as described in {{construct_manifest_primary}}), decodes the manifest, and determines the unique vehicle identifier. Additionally, the Director MAY utilize other mechanisms to uniquely identify a vehicle (e.g., 2-way TLS with unique client certificates).

Reviewer: ```suggestion
1. The Director SHOULD first identify the vehicle. This MAY be done when the Director receives a vehicle version manifest sent by a primary (as described in {{construct_manifest_primary}}), decodes the manifest, and determines the unique vehicle identifier. Additionally, the Director MAY utilize other mechanisms to uniquely identify a vehicle (e.g., 2-way TLS with unique client certificates).
```

Author: @trishankatdatadog - Do we want this first step to be conditional? 
Without identifying the vehicle the remainder of the steps do not seem possible.
I will change it to SHOULD per your suggestion to progress the PR, but think we should discuss further if it should be SHALL.

1. Using the vehicle identifier, the Director queries its inventory database (as described in {{inventory_db}}) for relevant information about each ECU in the vehicle.

> 2. The Director SHALL check the manifest for accuracy compared to the information in the inventory database. If any of the required checks fail, the Director MAY drop the request. An implementor MAY make whatever additional checks they wish. At a minimum, the Director SHALL check the following:

Reviewer 2: Best to leave the numbering alone since markdown can handle it automatically. Just leave every line at "1."

Reviewer 2: `SHALL drop`, right? Required checks are required.

Reviewer 3: Director might choose to take remediation steps outside the scope of Uptane in some cases, rather than just dropping the request. Could argue for a SHOULD instead, though.

Author: Thanks for the comments Patrick & Jon

Based on these ^ comments, I think we all agree to leave this second part as conditional.

A Director repository MUST conform to the following six-step process for directing the installation of software images on a vehicle.

> 1. The Director must first identify the vehicle. This MAY be done when the Director receives a vehicle version manifest sent by a primary (as described in {{construct_manifest_primary}}), decoding the manifest, and determine the unique vehicle identifier. Additionally, the Director MAY utilize other mechanisms to uniquely identify a vehicle (e.g., 2-way TLS with unique client certificates, RXSWIN, etc.).

Reviewer: ```suggestion
1. The Director must first identify the vehicle. This MAY be done when the Director receives a vehicle version manifest sent by a primary (as described in {{construct_manifest_primary}}), decodes the manifest, and determines the unique vehicle identifier. Additionally, the Director MAY utilize other mechanisms to uniquely identify a vehicle (e.g., 2-way TLS with unique client certificates, RXSWIN, etc.).
```

1. Using the vehicle identifier, the Director queries its inventory database (as described in {{inventory_db}}) for relevant information about each ECU in the vehicle.

> 1. The Director checks the manifest for accuracy compared to the information in the inventory database. If any of the required checks fail, the Director drops the request. An implementor MAY make whatever additional checks they wish. At a minimum, the Director SHALL check the following:

Reviewer: ```suggestion
2. The Director SHALL check that the manifest for accuracy compared to the information in the inventory database. If any of the required checks fail, the Director MAY drop the request. An implementor MAY make whatever additional checks they wish. At a minimum, the Director SHALL check the following:
```

Each ECU in a vehicle receiving over-the-air updates is either a primary or a secondary ECU. A primary ECU collects and delivers to the Director vehicle manifests ({{vehicle_version_manifest}}) containing information about which images have been installed on ECUs in the vehicle. It also verifies the time and downloads and verifies the latest metadata and images for itself and for its secondaries. A secondary ECU verifies the time and downloads and verifies the latest metadata and images for itself from its associated primary ECU. It also sends signed information about its installed images to its associated primary.

> All ECUs MUST verify image metadata as specified in {{metadata_verification}} before installing an image or making it available to other ECUs. A primary ECU MUST perform full verification ({{full_verification}}). A secondary ECU SHOULD perform full verification if possible. If a secondary cannot perform full verification, it SHALL perform partial verification. See [Uptane Deployment Considerations](#DEPLOY) for a discussion of how to choose between partial and full verification.

Reviewer 2: This should also be at most a `SHOULD` because not all secondaries are capable of performing any verification.

Author: Then based on my understanding of Uptane and by definition, this ECU is not an Uptane secondary.
I think we should have further discussion on what constitutes as a Secondary if you believe this should be conditional.

Reviewer: Are we on agreement on this? I think the understanding based on the last call was that we do not discuss ECUs that cannot do Uptane.

#### Send latest time to secondaries {#send_time_primary}

> The primary SHALL send the time to each ECU. The secondary will verify the time message, then overwrite its current time with the received time. The primary MAY omit this step if the secondary has its own way of loading and verifying the time, or cannot verify the time has detailed here.

Reviewer: ```suggestion
The primary SHALL send the time to each ECU. The secondary will verify the time message, then overwrite its current time with the received time. The primary MAY omit this step if the secondary has its own way of loading and verifying the time, or cannot verify the time as detailed here.
```

Reviewer 2: I don't understand why we would use `SHALL` if we the immediately say "actually you can skip this" immediately afterwards. Sounds like a `SHOULD` to me.

Author: Thanks for the comment, Patrick.

My thought is: if you make this a 'SHOULD' then this whole paragraph becomes conditional.
Therefore, there is *no* requirement for the primary to send the time to each ECU.
I believe this is more problematic than the current state.

Reviewer 3: Two things:

1. The caveat that the primary MAY omit the step if the secondary isn't capable of verifying the time is a catch-all. It is, most definitely, a SHOULD.
2. It would perhaps be slightly more proper to use the RFC2119 keyword CONDITIONALLY REQUIRED. Something like "Unless the secondary ECU has its own way of verifying the time, or does not have the capacity to verify a time message, the primary is CONDITIONALLY REQUIRED to send the time to each ECU. The secondary will verify the time message, then overwrite its current time with the received time."

Author: Updated per your recommendation.

#### Send latest time to secondaries {#send_time_primary}

> The primary SHALL send the time to each ECU. The secondary will verify the time message, then overwrite its current time with the received time. The primary MAY omit this step if the secondary has its own way of loading and verifying the time.

Reviewer 2: Should not be `SHALL` as some ECUs cannot support checks like that. `SHOULD` is appropriate here.

Author: OK that is fair. 

However, if it is optional then implementer's _may_ not send time to secondary ECUs which **are** capable of these checks.
I will rewrite the MAY sentence to indicate this step may be omitted for secondaries not capable of performing this check.

### Installing images on primary or secondary ECUs

An ECU SHALL perform the following steps when attempting to install a new image:

1. Verify latest attested time ({{verify_time}})

> 1. Verify metadata ({{verify_metadata}})

Reviewer: ```suggestion
2. Verify metadata ({{verify_metadata}})
```

> 1. Download latest image ({{download_image}})

Reviewer: ```suggestion
3. Download latest image ({{download_image}})
```

> 1. Verify image ({{verify_image}})

Reviewer: ```suggestion
4. Verify image ({{verify_image}})
```

> 1. Install image ({{install_image}})

Reviewer: ```suggestion
5. Install image ({{install_image}})
```

> 1. Create and send version report ({{create_version_report}})

Reviewer: ```suggestion
6. Create and send version report ({{create_version_report}})
```

Reviewer 4: As commented in other PR, it's better to avoid explicit numbers to simplify further changes - markdown renderer will take care of this automatically.

1. Load the Targets metadata file from the Director repository.
2. Find the Targets metadata associated with this ECU identifier.
3. Construct the Image filename using the rule in {{metadata_filename_rules}}, or use the download URL specified in the Director metadata.

> 4. If there is no Targets metadata about this image, abort the update cycle and report that there is no such image. Additionally, in the case of failure, the ECU SHALL overwrite its Targets metadata file with its previous Targets metadata file. Otherwise, download the image (up to the number of bytes specified in the Targets metadata), and verify that its hashes match the Targets metadata.

Reviewer: Seems oddly specific here. Shouldn't this generally be the case for any metadata verification failure at all? @JustinCappos @awwad That is, no new metadata should replace previous metadata until all of them are verified.

Also, we should not mention files directly, as not all implementors would use files.

Reviewer 5: ```suggestion
4. If there is no Targets metadata about this image, abort the update cycle and report that there is no such image. Additionally, in the case of failure, the ECU SHALL retain its previous Targets metadata instead of using the new Targets metadata. Otherwise, download the image (up to the number of bytes specified in the Targets metadata), and verify that its hashes match the Targets metadata.
```

If the ECU has secondary storage, the checks SHOULD be performed on the image in secondary storage, before it is installed.

> If any step fails, the ECU SHALL jump to the ({{create_version_report}}) step. If an ECU does not have secondary storage, then before attempting verification, the ECU SHALL ensure a backup of a previous working image and associated metadata is created. This MAY be stored on the updating ECU or elsewhere (e.g., associated primary). Thus, if a step fails and the ECU does not have secondary storage, the ECU SHALL install the backup image to ensure it is in a working condition.

Reviewer: Sorry, but this doesn't make sense. If the ECU does _not_ have secondary storage, how SHALL it ensure a backup of the previous working image on itself, if not elsewhere?

Reviewer: Also, which step should fail before the ECU installs the backup image? What if the new image hasn't been replaced yet?

Reviewer 6: I have concerns about the ECU's ability to ensure a backup is there and valid if it is not doing it itself. How would it ensure the integrity of the image or that it was stored properly?

Author: Thanks for the comments, Phil.

I agree this is a difficult problem to solve.
Overall, I think everyone can agree that the vehicle needs to be in a 'functioning state' in case of update failure.
I am not saying the approach I have outlined above is optimal, but at least it starts a conversation on how to ensure a fail-back mode is possible for partial-verification (pv) secondaries which do not contain additional storage for previous updates (i.e., A/B partition not available).

My approach was to ensure a couple of important things happened: 
    1. Backup the currently working image & metadata before attempting update
    [Failure occurs during update process]
    2. Update pv-secondary to the previously working version 
    3. Report this failure to the server (not currently in PR, but probably should be expanded to include)

To answer your question, Phil, maybe we can be more verbose in a couple of areas.
For example, in Step 1^, document that the storage location (e.g., Primary), has notified the pv-secondary that storage was successfully completed, before pv-secondary begins update process.
For ensuring the integrity, maybe the approach we take is requiring pv-secondaries to store a hash of its previously working image, which is not overwritten until a successful update has occurred. 
The pv-secondary can use this stored hash to verify against hash in metadata & image received from associated update provider (e.g., primary).

These ^^ don't have to be the concrete solution or wording we put into the standard, but I do hope this facilitates additional feedback which we can include in v1.0 to ensure the standard accounts for failure cases.

> If any step fails, the ECU SHALL jump to the ({{create_version_report}}) step. If an ECU does not have secondary storage, then before attempting verification, the ECU SHALL ensure a backup of a previous working image and associated metadata is created. This MAY be stored on the updating ECU or elsewhere (e.g., associated primary). Thus, if a step fails and the ECU does not have secondary storage, the ECU SHALL install the backup image to ensure it is in a working condition.

Reviewer: ```suggestion
If any step fails, the ECU SHALL jump to the ({{create_version_report}}) step. If an ECU does not have secondary storage, then before attempting verification, the ECU SHOULD ensure a backup of a previous working image and associated metadata is created. This MAY be stored on the updating ECU or elsewhere (e.g., associated primary). Thus, if a step fails and the ECU does not have secondary storage, the ECU SHOULD install the backup image to ensure it is in a working condition.
```

> If any step fails, the ECU SHALL jump to the ({{create_version_report}}) step. If an ECU does not have secondary storage, then before attempting verification, the ECU SHALL ensure a backup of a previous working image and associated metadata is created. This MAY be stored on the elsewhere within the vehicle (e.g., associated primary). Thus, if a step fails and the ECU does not have secondary storage, the ECU SHALL install the backup image to ensure it is in a working condition.

Reviewer 2: This is actually a bit confusing (and was before this change). This verification is done before installation, so why would we need to re-install the backup image if we haven't installed the bogus one yet? I suppose the preceding line does say "SHOULD be performed ... before it is installed" so there is a possibility it was installed before verification.

Author: Thanks for the feedback, Patrick.

I agree my PR could be revised to be more cohesive with the rest of the section. (please see updated version)
The specific use case I am addressing is *pv-secondaries with no additional storage*.
This means in Step 1 of {#verify_image}, they have *already overwritten* the Targets metadata they previously contained.
Hence, any failure during these steps indicate they have overwritten their previously valid metadata (and potentially image).

I have pushed a new change to address this confusion.
Please let me know what you think.

Reviewer 2: Looks better to me, thanks!

If the ECU has secondary storage, the checks SHOULD be performed on the image in secondary storage, before it is installed.

> If any step fails, the ECU SHALL jump to the fifth step ({{create_version_report}}). If a step fails and the ECU does not have secondary storage, then the ECU SHALL ensure a backup of a previous working image is stored before attempting the verification. Upon failure, the ECU SHALL install the backup image to ensure it is in a working condition.

Reviewer: ```suggestion
If any step fails, the ECU SHALL jump to the ({{create_version_report}}) step. If a step fails and the ECU does not have secondary storage, then the ECU SHOULD ensure a backup of a previous working image is stored before attempting the verification. Upon failure, the ECU SHOULD install the backup image to ensure it is in a working condition.
```

Reviewer: We cannot force secondary storage on all implementations, therefore the SHALL -> SHOULD

Author: I agree with not forcing all ECUs to have secondary storage, sorry for my confusion.
My intent was to force implementor's to have a backup of current running image & associated metadata in case of update failure.
This backup can be stored on the updating ECU, on another ECU with sufficient storage, on a dealer tool, etc.

Does this clarify my intent?
I will provide some clarity in my wording...

#### Install image {#install_image}

> The ECU SHALL attempt to install the update at a time when all pre-conditions are met. These pre-conditions MAY include ensuring the vehicle is in a safe environment for install (e.g., vehicle is in Park when updating a specific ECU). Another pre-condition MAY include ensuring the ECU has a backup of its current image and metadata in case the current installation fails.

Reviewer: ```suggestion
The ECU SHOULD attempt to install the update at a time when all pre-conditions are met. These pre-conditions MAY include ensuring the vehicle is in a safe environment for install (e.g., vehicle is in the parked state when updating a specific ECU). Another pre-condition MAY include ensuring the ECU has a backup of its current image and metadata in case the current installation fails.
```

Reviewer: SHALL should be reserved for security relevant to Uptane, not necessarily the vehicle safety, which is outside its scope

Author: To avoid making the entire section conditional, I rewrote it to fit your recommendation.
I think we can have another discussion on the relationship between vehicle safety & cybersecurity, but we can save that until the 14th :)

#### Create and send version report {#create_version_report}

The ECU SHALL create a version report as described in {{version_report}}, and send it to the primary (or simply save it to disk, if the ECU is a primary). The primary SHOULD write the version reports it receives to disk and associate them with the secondaries that sent them.

### Metadata verification procedures {#metadata_verification}

> A primary ECU MUST perform full verification of metadata. A secondary ECU SHOULD perform full verification of metadata. If a secondary cannot perform full verification it SHALL perform partial verification instead.

Reviewer 2: This should be at most `SHOULD` as well.

Author: Then based on my understanding of Uptane and by definition, this ECU is not an Uptane secondary.
I think we should have further discussion on what constitutes as a Secondary if you believe this should be conditional.

Reviewer: See the comment above

11. Verify that Targets metadata from the Director and Image repositories match. A primary ECU MUST perform this check on metadata for all images listed in the Targets metadata file from the Director repository downloaded in step 6. A secondary ECU MAY elect to perform this check only on the metadata for the image it will install. (That is, the target metadata from the Director that contains the ECU identifier of the current ECU.) To check that the metadata for an image matches, complete the following procedure:
    1. Locate and download a Targets metadata file from the Image repository that contains an image with exactly the same file name listed in the Director metadata, following the procedure in {{resolve_delegations}}.
    2. Check that the Targets metadata from the Image repository matches the Targets metadata from the Director repository:

>         1. Check that the non-custom metadata (i.e., length and hashes) of the unencrypted or encrypted image are the same in both sets of metadata. Note: the primary is responsible for validating encrypted images and associated metadata. The target ECU (primary or secondary) is responsible for validating the unencrypted image and associated metadata.

Reviewer 2: Again, I don't see how this is possible. The images repository doesn't know anything about encrypted images; they only live in the Director repository. It would be quite difficult to get that metadata to match in that case.

Author: Thanks for the comment, Patrick.

This was discussed in-length on a call 1 or 2 weeks ago?
Essentially, the comment "image repository does not have encrypted images" is one that will not be 100% true in production.
Ira gave a very good example of companies wanting to protect IP and ensuring images are encrypted until they reach their **target** ECU.
To account for this, I put this language in the PR.
If you have a proposal for better language to include, please let me know.

Reviewer 2: In principle, I have no objection to the language here. If others agree that it doesn't conflict with the encryption language described elsewhere, then I have no problem with it.

Reviewer 6: Supporting encrypted images is certainly necessary in production. For the encrypted data usecase, verification of unencrypted data may only be possible on the target ECU due to knowledge of the encryption keys. One very important consideration is the order of operations for encryption / verification. We need to be very clear in what steps the encryption happens. In my opinion, we need to require validation of the unencrypted data on the target ECU to ensure end to end authenticity and integrity.

11. Verify that Targets metadata from the Director and Image repositories match. A primary ECU MUST perform this check on metadata for all images listed in the Targets metadata file from the Director repository downloaded in step 6. A secondary ECU MAY elect to perform this check only on the metadata for the image it will install. (That is, the target metadata from the Director that contains the ECU identifier of the current ECU.) To check that the metadata for an image matches, complete the following procedure:
    1. Locate and download a Targets metadata file from the Image repository that contains an image with exactly the same file name listed in the Director metadata, following the procedure in {{resolve_delegations}}.
    2. Check that the Targets metadata from the Image repository matches the Targets metadata from the Director repository:
        1. Check that the non-custom metadata (i.e., length and hashes) of the unencrypted or encrypted image are the same in both sets of metadata.

> 	    - Please note, the primary is responsible for validating encrypted images and associated metadata. The target ECU (primary or secondary) is responsible for validating the unencrypted image and associated metadata.

Reviewer 3: ```suggestion
            - Please note, the primary is responsible for validating encrypted images and associated metadata. The target ECU (primary or secondary) is responsible for validating the unencrypted image and associated metadata.
```

> 	        1. Please note, the primary is responsible for validating encrypted images and associated metadata. The target ECU (primary or secondary) is responsible for validating the unencrypted image and associated metadata.

Reviewer 3: No tabs, please--it's undefined behaviour in kramdown-rfc2629.

Author: fixed - please let me know if error still persists.

Reviewer 3: There's still a tab at the beginning of line 837.

>             - Please note, the primary is responsible for validating encrypted images and associated metadata. The target ECU (primary or secondary) is responsible for validating the unencrypted image and associated metadata.

Reviewer: ```suggestion
            - Note, the primary is responsible for validating encrypted images and associated metadata. The target ECU (primary or secondary) is responsible for validating the unencrypted image and associated metadata.
```

Author: Updated.

Also, allow me to respond to your question from lines 644&805 here. 

> "Are we on agreement on this? I think the understanding based on the last call was that we do not discuss ECUs that cannot do Uptane."

I agree this was the result of the conversation from the last call.
The way I interpreted this was to include this explicit detail, rather than the previous approach which made it seem more conditional for a secondary to perform verification.

Hope this clarification helps!

Reviewer: Could you please make a quick fix? Please say `Note: ` instead of `Note, `

Reviewer: Also remove the extra indentation, please. We can just include it at the end of the sentence preceding right before.

11. Verify that Targets metadata from the Director and Image repositories match. A primary ECU MUST perform this check on metadata for all images listed in the Targets metadata file from the Director repository downloaded in step 6. A secondary ECU MAY elect to perform this check only on the metadata for the image it will install. (That is, the target metadata from the Director that contains the ECU identifier of the current ECU.) To check that the metadata for an image matches, complete the following procedure:
    1. Locate and download a Targets metadata file from the Image repository that contains an image with exactly the same file name listed in the Director metadata, following the procedure in {{resolve_delegations}}.
    2. Check that the Targets metadata from the Image repository matches the Targets metadata from the Director repository:

>         1. Check that the non-custom metadata (i.e., length and hashes) of the unencrypted or encrypted image are the same in both sets of metadata.

Reviewer 2: The image repository does not know anything about encrypted images, so it cannot have metadata about them.

Author: I think it is unrealistic to force an implementer to have unencrypted images on the repository.

I think this is an issue we should gain alignment on.

Author: I looked a little closer a this section/change.
This is included in the process for a Primary to verify metadata.
It states that the Primary shall check the Targest metadata from Image repo matches Director Repo [838].
This is further expanded upon to check the non-custom metadata, including length & hash on unencrypted image.
In the case that the image is encrypted with a key that **only** a specific secondary can decrypt, the primary will fail at this step.
Hence, my clarification above.

I will add: "the primary will validate encrypted images & metadata, but target ECU must validate the unencrypted image and its associated metadata."

3. Check that it has been signed by the threshold of keys specified in the relevant metadata file (Checks for an arbitrary software attack):
    1. If checking top-level targets metadata, the threshold of keys is specified in the Root metadata.
    2. If checking delegated targets metadata, the threshold of keys is specified in the targets metadata file that delegated authority to this role.
4. Validate the signatures of the Targets metadata.
	1. If checking top-level targets metadata signatures, the public keys are specified in the Root metadata.

> 	2. If checking delegated targets metadata signatures, the public keys are specified in the targets metadata file that delegated authority to this role.

Reviewer 2: I agree with @tkfu that this is redundant with the preceding step.

Author: Fair.
I removed and added clarity to previous step.
